book/src/leader-validator-transition.md

```diff
@@ -0,0 +1,157 @@
+# Leader and Validator Transition
+
+A fullnode operates in two modes, leader and validator.  The modes overlap in
```

Comment:
I wonder if this decision decision, of saying a fullnode operates in one of two modes, is the source of a lot of confusion.  Perhaps we should call these things validators that sometimes hold the role of *slot leader*, which just means it gets its transactions from a different source and processes them slightly differently.  As a slot leader, they'd pull transactions off the TPU port instead of pulling entries off the TVU port. Likewise, after processing the transactions, if any produce an error, a leader should filter out the bad, whereas the validator should reject the batch. After processing, the leader should wrap the transactions in an entry. Afterward, either the leader or validator should lock down the new state by registering the PoH hash with the bank.

Reply:
@garious 

Yea, easier said than done :).  How does a running replay stage get shut down and stop voting when a leader starts producing data? 

Another idea i was thinking of is separating the voting stage from the replay stage.  So both the tpu and the tvu can notify it.  But the TPU needs an on/off switch toggled by the PoH entries that connect it to some previous voted block.

Reply:
the replay stage needs to keep running in case this node is behind the network...

Reply:
@rob-solana, how would it get into that state in this loop?

Reply:
@aeyakovenko the case I'm considering is the one where the node starts a TPU when in fact the rest of the network has already decided to skip the node's transmission.  This can occur during a temporary network outage.  The result is that this node becomes even further behind, having disabled its TVU during its leader slot.

Reply:
@rob-solana, the way the loop is written is that the TPU proposes a slot, but the toy will skip voting for that slot if a better fork option exists

Reply:
right, but only if the TVU is running, or not until the TVU starts up again

Reply:
If the node has enough compute to run both TPU and TVU concurrently, than the node has enough compute to catch up :)

Reply:
hm, if par_process_entries() is working right...  ok.  I'm happy.